File: agentcore/src/main/java/agentcore/gui/SpringFxmlLoader.java
package agentcore.gui;

import javafx.fxml.FXMLLoader;
import javafx.scene.Parent;
import org.springframework.beans.BeansException;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.io.InputStream;

/**
 * @author Nikita Gorodilov
 */
@Component
public class SpringFxmlLoader implements ApplicationContextAware {

    private ApplicationContext applicationContext;

    public Parent load(InputStream inputStream) {
        FXMLLoader loader = new FXMLLoader();
        loader.setControllerFactory(applicationContext::getBean);

        try {
            return loader.load(inputStream);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
        this.applicationContext = applicationContext;
    }
}

File: agentcore/src/main/java/agentcore/gui/SpringFxmlLoader.kt
package agentcore.gui

import javafx.fxml.FXMLLoader
import javafx.scene.Parent
import javafx.util.Callback
import org.springframework.beans.BeansException
import org.springframework.context.ApplicationContext
import org.springframework.context.ApplicationContextAware
import org.springframework.stereotype.Component

import java.io.IOException
import java.io.InputStream

/**
 * @author Nikita Gorodilov
 */
@Component
open class SpringFxmlLoader : ApplicationContextAware {

    private var applicationContext: ApplicationContext? = null

    fun load(inputStream: InputStream): Parent {
        val loader = FXMLLoader()
        loader.controllerFactory = Callback<Class<*>, Any> { applicationContext!!.getBean(it) }

        try {
            return loader.load<Parent>(inputStream)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    @Throws(BeansException::class)
    override fun setApplicationContext(applicationContext: ApplicationContext) {
        this.applicationContext = applicationContext
    }
}
